docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,229 @@
+# [RFC] Conjure Endpoint Errors
+
+**Date:** 2020/03/19
+
+**Stakeholders:**
+* *Anyone that produces or consumes conjure-generated code.*
+
+* * *
+
+## **Context**
+
+Currently, errors can be defined in Conjure type definitions. These errors can then be thrown by services, and if a client 
+knows about them, they can be handled during runtime by clients. The exact handling of this mechanism may vary by
+the implementation (for example, in Java these are `ServicException`s, which become `RemoteException`s which can be caught
+by the client).
+
+There are multiple issues with this. As a consumer, you only know that an endpoint may throw an error by reading the 
+documentation or the implementation itself. As a producer, you are limited in the expressivity of what you can say 
+with your exception, as you have no guarantee that a consumer will handle your exception (or even read your documentation).
+
+This RFC introduces *declared errors* on Conjure endpoints. These errors represent real outcomes of a call to an endpoint, 
+and should be handled by consumers.
+
+## Existing Documentation
+
+### PRs for example implementation (in Java, using checked exceptions)
+
+* Root PR (Conjure API changes): https://github.com/palantir/conjure/pull/816
+* Conjure Java API Changes: https://github.com/palantir/conjure-java-runtime-api/pull/634
+* Conjure Java & Dialogue implementation: https://github.com/palantir/conjure-java/pull/1275
+
+## Example Conjure and Java implementations
+
+**Conjure API**
+
+```
+types:
+  definitions:
+    default-package: com.palantir.my.api
+    errors:
+      BlueMoon:
+        namespace: MyConjureService
+        code: CUSTOM_SERVER
+
+services:
+  MyConjureService:
+    name: An example service.
+    package: com.palantir.my.api
+    default-auth: header
+    endpoints:
+      getSomeResult:
+        http: GET /
+        returns: list<rid>
+        errors:
+          - BlueMoon
```

Comment:
What assumptions about completeness/exhaustiveness should users be able to make here?

Clearly in Java, runtime exceptions are _much_ more ergonomic to use (i.e. in streams/lambdas), so it seems like people will not adopt checked exceptions _everywhere_ in backends.

In that case, I worry that someone will introduce a new conjure error type and start throwing it deep inside one of their stores or implementation detail classes, not realizing that there are actually _multiple_ endpoints which call this and could now start throwing these errors?

Reply:
> In that case, I worry that someone will introduce a new conjure error type and start throwing it deep inside one of their stores or implementation detail classes, not realizing that there are actually multiple endpoints which call this and could now start throwing these errors?

Checked exceptions are validated by the type system, attempting to throw a checked exception from an endpoint that doesn't declare one will fail to compile. I'd expect this to result in the api definition being updated to better reflect possible failures for each endpoint. You're correct that the exception may be wrapped with an unexpected runtime exception, but isn't that true of runtime exceptions as well?

Reply:
Note that the generator doesn't need to use checked exceptions. We could generate `Result<...>` types which are very stream/lambda friendly.

Reply:
One issue that's come up with Result types in Java code is that developers have to remember to inspect the result and convert the error state to an exception. Not a deal breaker, (and something we can handle with errorprone) but worth noting

Reply:
We can't make any claims about exhaustiveness because there's no guarantee that people stay up to date with the generated client bindings.

Reply:
I'd personally vote against result types, as they aren't idiomatic java. Basically, I think it would be nice if client bindings for each language could approximate idiomatic error handling in that language, e.g.:

* Java - checked (or unchecked) exceptions
* C#/python - unchecked exceptions
* golang/rust - (result, error) tuples w/ discriminated unions on the error type

Reply:
From a Typescript/Javascript perspective, only unchecked exceptions really make sense as exception handling is very rare. 

I would want to get some feedback from other FE devs but I would probably suggest revving major version of generated libraries if we were to do this so I could write code like:

```ts
const r = service. getSomeResult();
if (r.result) {
  doStuffWithListOfRids(r.result);
} else {
  // treat r.exception as a conjure union of exceptions so unknown can be handled.
}